keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,877 @@
+
+### Goals
+
+#### Phase 1
+
+- Users can specify an optional list of commands in `Pod`s specification at container level.
+- Users or upper level controllers can request to execute commands in `Pod`s by creation of an API object(`PodNotification`).
+- Implement a *trusted* controller which monitors requests to execute commands, notifies the corresponding `Pod`s to run those commands, and updates status of requests corresponding to the execution results.
+
+#### Phase 2
+
+- Introduce an upper level API(`Notification`) which allows users to send `PodNotification` to a set of `Pod`s which can be matched by a label selector.
+- Extend `ContainerNotifier` API in Phase 1 to support sending signals to containers.
+- Move *trusted* controller logic into Kubelet.
+
+#### Phase 3
+
+- Add a `Probe` to verify the results from the commands/signals if needed.
+
+### Non-Goals
+
+- Writing concrete commands, for example quiesce scripts, which could be executed. It is the responsibility of users who set up pod definitions.
+
+## Proposal
+
+In phase 1, propose to introduce several API changes:
+- Adding an optional field `Notifiers` which is a list of `ContainerNotifier`s into `Container`.
+- Adding an inlined type `ContainerNotifierHandler` which defines a command.
+- Adding a core API type `PodNotification` which defines request type to trigger execution of `ContainerNotifier`(s) in a `Pod`.
+- Introduce a new gate `ContainerNotifier` to toggle this feature.
+
+A SINGLE *trusted* controller(Pod Notification Controller) will be implemented to watch `PodNotification` resources, execute the command and update their statuses accordingly.
+
+In phase 2, propose to:
+- Add a core API `Notification` type and a controller which processes `Notification` resources.
+- Add an inline `Pod` definition for signals and allows the API object to send a request to trigger delivering of those signals.
+- Move logic in the Pod Notification controller into Kubelet. Kubelet watches `PodNotification` objects, runs the command and updates statuses of `PodNotification` objects accordingly.
+
+In phase 3, a `Probe` may be added if needed as an inline `Pod` definition to verify the signal is delivered or whether the command is run and results in the desired outcome.
+
+Details of proposed API changes can be found in *API Changes* section  bellow.
+
+### API Changes
+
+#### Phase 1 API Changes
+
+##### Inline Definition - ContainerNotifier
+
+Add []ContainerNotifier to the Container struct:
+
+```
+type Container struct {
+    ......
+    // +optional
+    Lifecycle *Lifecycle `json:"lifecycle,omitempty" protobuf:"bytes,12,opt,name=lifecycle"`
+
+    // Notifiers can be triggered by PodNotification resources.
+    // Each notifier must have a unique name within the Container.
+    // Upon creation of a PodNotification, Container(s) in the specified
+    // Pod with a matching notifier name will be executed.
+    // +optional
+    Notifiers []ContainerNotifier
```

Comment:
Is this mutable? Should someone be able to add a notifier to a pod after it is running I guess is the question that I am asking?

Reply:
"Containers" field in Pod can not be updated that is my understanding.
[ref](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core)
IMO, it also seems to be reasonable to make the "notifiers" immutable such that only a restricted/predefined set of notifiers can be triggered. Are there any specific cases you see the needs of dynamically adding/removing notifiers from a specific container?

Reply:
I was wondering if it would be possible to do that, I can't think of a good use case thanks!